### PR TITLE
[FSDP] Ensure that customized non tensor optimizer state can be saved

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99214

The current logic does not actually handle all different non-tensor optimizer states correctly. This PR fixes the issue and adds a test.

This PR will solve https://github.com/pytorch/pytorch/issues/99079

Differential Revision: [D45021331](https://our.internmc.facebook.com/intern/diff/D45021331/)

### CODE REVIEW

File: test/distributed/fsdp/test_fsdp_optim_state.py [hunk 1]
```diff
@@ -1807,6 +1807,33 @@ def step():
...
+            original_osd = deepcopy(optim.state_dict())
+            for param_id, state in original_osd["state"].items():
+                # Addd customized value
+                state["value1"] = 2.74
+                state["value2"] = None
+
+            osd = FSDP.optim_state_dict(model, optim, optim_state_dict=original_osd)
+            osd_to_load = FSDP.optim_state_dict_to_load(model, optim, osd )
+            for param_id, state in osd_to_load["state"].items():
+                # Addd customized value
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_optim_state.py:1831]: nit: typo 😄 
```suggestion
                # Add customized value
```

File: test/distributed/fsdp/test_fsdp_optim_state.py [hunk 2]
```diff
@@ -1807,6 +1807,33 @@ def step():
...
+
+            def step():
+                loss = model(model.get_input())
+                loss.backward(loss)
+                optim.step()
+
+            step()
+            original_osd = deepcopy(optim.state_dict())
+            for param_id, state in original_osd["state"].items():
+                # Addd customized value
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_optim_state.py:1824]:
```suggestion
                # Add customized value
```

File: test/distributed/fsdp/test_fsdp_optim_state.py [hunk 3]
```diff
@@ -1807,6 +1807,33 @@ def step():
...
+                state["value1"] = 2.74
+                state["value2"] = None
+
+            osd = FSDP.optim_state_dict(model, optim, optim_state_dict=original_osd)
+            osd_to_load = FSDP.optim_state_dict_to_load(model, optim, osd )
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_optim_state.py:1829]: nit: I wonder if lint will complain about this one
```suggestion
            osd_to_load = FSDP.optim_state_dict_to_load(model, optim, osd)
```

**Review by awgu — APPROVED** [review]
Thanks for the super fast fix! This looks good to me. I just left a few nits.

**Review by awaelchli — APPROVED** [review]
Thanks 🎉 !

**Pull request closed without merging**